File: 2016_11_07/quicksort/quicksort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int partition(int a[], int len, int lo, int hi) {
	int p = a[lo];
	int i = lo;
	int j = hi;
	while (i < j) {
		while (i < len && a[i] <= p) {
			i++;
		}
		while (a[j] > p) {
			j--;
		}
		if (i < j) {
			int t = a[i];
			a[i] = a[j];
			a[j] = t;
		}
	}
	a[lo] = a[j];
	a[j] = p;
	return j;
}
void quicksort(int a[], int len, int lo, int hi) {
	if (lo < hi) {
		int p = partition(a, len, lo, hi);
		quicksort(a, len, lo, p-1);
		quicksort(a, len, p+1, hi);
	}
}

int main(int argc, char* argv[]) {
	int size; sscanf(argv[1], "%d", &size);
	int test[size];
	for (int i = 0; i < size; i++) {
		test[i] = i;
	}
	srand(time(NULL));
	for (int i = 0; i < size; i++) {
		int r = rand() % (size - i) + i;
		int t = test[i];
		test[i] = test[r];
		test[r] = t;
	}

	if (size < 32) {
		printf("[");
		for (int i = 0; i < size-1; i++) {
			printf("%d, ", test[i]);
		}
		printf("%d]\n", test[size-1]);
	}

	clock_t begin = clock();
	quicksort(test, size, 0, size-1);
	clock_t end = clock();

	if (size < 32) {
		printf("[");
		for (int i = 0; i < size-1; i++) {
			printf("%d, ", test[i]);
		}
		printf("%d]\n", test[size-1]);
	}

	clock_t elapsed = end - begin;

	printf("Begin: %d µs\n", begin);
	printf("End: %d µs\n", end);
	printf("Time: %d µs\n", elapsed);
}

File: 2016_11_24/quicksort/quicksort.py
#!/usr/bin/python3

import random, sys, time

def pivot(a, lo, hi):
	# Median of Three
	b = random.randrange(lo, hi+1)
	c = random.randrange(lo, hi+1)
	d = random.randrange(lo, hi+1)
	if a[b] < a[c]:
		if a[c] < a[d]:
			return c
		else:
			if a[b] < a[d]:
				return d
			else:
				return b
	else:
		if a[b] < a[d]:
			return b
		else:
			if a[c] < a[d]:
				return d
			else:
				return c

def partition(a, lo, hi):
	x = pivot(a, lo, hi)
	a[x], a[lo] = a[lo], a[x]
	p = a[lo]
	i = lo
	j = hi
	while i < j:
		while i < len(a) and a[i] <= p:
			i += 1
		while a[j] > p:
			j -= 1
		if i < j:
			a[i], a[j] = a[j], a[i]
	a[lo], a[j] = a[j], a[lo]
	return j

def quicksort(a, lo, hi):
	if lo < hi:
		p = partition(a, lo, hi)
		quicksort(a, lo, p-1)
		quicksort(a, p+1, hi)

def validate(a):
	for i in range(1, len(a)):
		if a[i] < a[i-1]:
			return False
	return True

size = int(sys.argv[1])
test = []
for i in range(size):
	test.append(random.randrange(size))

print("===============================================================================")
if size < 32: print(test)
begin = time.time()
quicksort(test, 0, size-1)
end = time.time()
if size < 32: print(test)
print("Is Valid:", validate(test))
print("===============================================================================")
print("Elapsed:", int(round((end - begin) * 1)), "s")
print("Elapsed:", int(round((end - begin) * 1000)), "ms")
print("Elapsed:", int(round((end - begin) * 1000000)), "ms")
print("Elapsed:", int(round((end - begin) * 1000000000)), "ns")
print("===============================================================================")
test = []
for i in range(size):
	test.append(i)
print("===============================================================================")
if size < 32: print(test)
begin = time.time()
quicksort(test, 0, size-1)
end = time.time()
if size < 32: print(test)
print("Is Valid:", validate(test))
print("===============================================================================")
print("Elapsed:", int(round((end - begin) * 1)), "s")
print("Elapsed:", int(round((end - begin) * 1000)), "ms")
print("Elapsed:", int(round((end - begin) * 1000000)), "ms")
print("Elapsed:", int(round((end - begin) * 1000000000)), "ns")
print("===============================================================================")
test = []
for i in range(size):
	test.append(size-i)
print("===============================================================================")
if size < 32: print(test)
begin = time.time()
quicksort(test, 0, size-1)
end = time.time()
if size < 32: print(test)
print("Is Valid:", validate(test))
print("===============================================================================")
print("Elapsed:", int(round((end - begin) * 1)), "s")
print("Elapsed:", int(round((end - begin) * 1000)), "ms")
print("Elapsed:", int(round((end - begin) * 1000000)), "ms")
print("Elapsed:", int(round((end - begin) * 1000000000)), "ns")
print("===============================================================================")









